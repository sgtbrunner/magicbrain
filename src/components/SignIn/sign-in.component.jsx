import React, { useState } from 'react';
import PropTypes from 'prop-types';

import api from '../../utils/api.utils';

const SignIn = ({ loadUser, onRouteChange, clearFields }) => {
  const [signInData, setSignInData] = useState({
    email: '',
    password: '',
  });

  const { email, password } = signInData;

  const onEmailChange = (event) => {
    setSignInData({ ...signInData, email: event.target.value.toLowerCase() });
  };

  const onPasswordChange = (event) => {
    setSignInData({ ...signInData, password: event.target.value });
  };

  const onSignInSubmit = () => {
    if (email && password) {
      api.signInUser({ email, password }).then((user) => {
        if (user?.id) {
          loadUser(user);
          onRouteChange('home');
        } else {
          window.alert('Invalid user and/or password');
          clearFields();
        }
      });
    } else {
      window.alert('Email and Password fields should be filled in');
    }
  };

  const handleKeyPress = (event) => {
    return event.key === 'Enter' ? onSignInSubmit() : null;
  };

  return (
    <article
      className="br3 ba b--black-10 mv4 w-100 w-50-m w-25-l mw6 shadow-5 center form smaller"
      onKeyPress={handleKeyPress}
    >
      <main className="pa4 black-80">
        <div className="measure">
          <fieldset id="sign_in" className="ba b--transparent ph0 mh0">
            <legend className="f2 fw6 ph0 mh0 noselect">Sign In</legend>
            <div className="mt3">
              <label className="db fw6 lh-copy f6" htmlFor="email-address">
                Email
              </label>
              <input
                className="pa2 input-reset ba bg-white w-100"
                type="email"
                name="email-address"
                id="email-address"
                onChange={onEmailChange}
                required
              />
            </div>
            <div className="mv3">
              <label className="db fw6 lh-copy f6" htmlFor="password">
                Password
              </label>
              <input
                className="pa2 input-reset ba bg-white w-100"
                type="password"
                name="password"
                id="password"
                onChange={onPasswordChange}
                required
              />
            </div>
          </fieldset>
          <div className="">
            <input
              className="b ph3 pv2 input-reset ba b--black bg-lightest-blue grow pointer f6 dib"
              type="submit"
              value="Sign in"
              onClick={onSignInSubmit}
            />
          </div>
          <div className="lh-copy mt3">
            <p className="f5 link dim black db pointer" onClick={() => onRouteChange('register')}>
              Register now, it's free!
            </p>
          </div>
        </div>
      </main>
    </article>
  );
};

SignIn.propTypes = {
  loadUser: PropTypes.func.isRequired,
  onRouteChange: PropTypes.func.isRequired,
  clearFields: PropTypes.func.isRequired,
};

export default SignIn;
