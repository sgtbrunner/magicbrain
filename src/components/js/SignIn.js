import React from 'react';

class SignIn extends React.Component {
  constructor() {
    super()
      this.state = {
        signInEmail: '',
        signInPassword: ''
      }
  }

  onEmailChange = (event) => {
    this.setState({signInEmail: event.target.value.toLowerCase()});
  }

  onPasswordChange = (event) => {
    this.setState({signInPassword: event.target.value})
  }

  handleKeyPress = (event) => {
    if(event.key === 'Enter'){
      this.onSignInSubmit();
    }
  }

  onSignInSubmit = () => {
    if(this.state.signInEmail && this.state.signInPassword) {
      fetch('http://localhost:3000/signin', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({
          "email": this.state.signInEmail,
          "password": this.state.signInPassword
        })
      })
      .then(response => response.json())
      .then(user => {
        if(user.id){
          this.props.loadUser(user);
          this.props.onRouteChange('home'); //As Register is a class, it automatically receives props from App.js and (props) are accessed through this.props.
        } else {
          window.alert('Invalid user and/or password');
          this.props.clearFields();
        }
      });
    } else {
      window.alert ('Email and Password fields should be filled in');
    }
  }

  render() {
    return (
      <article className="br3 ba b--black-10 mv4 w-100 w-50-m w-25-l mw6 shadow-5 center form smaller"
               onKeyPress={this.handleKeyPress} 
      >
          <main className="pa4 black-80">
              <div className="measure">
                  <fieldset id="sign_in" className="ba b--transparent ph0 mh0">
                    <legend className="f2 fw6 ph0 mh0 noselect">Sign In</legend>
                      <div className="mt3">
                        <label className="db fw6 lh-copy f6" htmlFor="email-address">Email</label>
                        <input 
                          className="pa2 input-reset ba bg-white w-100" 
                          type="email" 
                          name="email-address"  
                          id="email-address"
                          onChange={this.onEmailChange}
                          required
                        />
                      </div>
                      <div className="mv3">
                        <label className="db fw6 lh-copy f6" htmlFor="password">Password</label>
                        <input 
                          className="pa2 input-reset ba bg-white w-100" 
                          type="password" 
                          name="password"  
                          id="password"
                          onChange={this.onPasswordChange}
                          required 
                        />
                      </div>
                  </fieldset>
                    <div className="">
                      <input
                        className="b ph3 pv2 input-reset ba b--black bg-lightest-blue grow pointer f6 dib" 
                        type="submit" 
                        value="Sign in"
                        onClick={this.onSignInSubmit}                 
                      />
                    </div>
                    <div className="lh-copy mt3">
                      <p className="f5 link dim black db pointer"
                         onClick={() => this.props.onRouteChange('register')} //As SignIn is a class, it automatically 
                         // receives props from App.js and (props) are accessed through this.props.
                      >
                        Register now, it's free!
                      </p>
                    </div>
              </div>
            </main>
      </article>
    );
  }
}

export default SignIn;