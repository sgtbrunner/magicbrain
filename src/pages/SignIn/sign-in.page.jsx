import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';

import api from '../../utils/api.utils';
import {
  INPUT_INITIAL_STATE,
  EMAIL_ERROR_MESSAGE,
  PASSWORD_ERROR_MESSAGE,
  EMAIL_REGEX_KEY,
} from '../../utils/constants.utils';

const SignIn = ({ loadUser }) => {
  const history = useHistory();

  const [email, setEmail] = useState({
    ...INPUT_INITIAL_STATE,
    errorText: EMAIL_ERROR_MESSAGE,
  });
  const [password, setPassword] = useState({
    ...INPUT_INITIAL_STATE,
    errorText: PASSWORD_ERROR_MESSAGE,
  });

  const getInput = {
    email,
    password,
  };

  const getSetInput = {
    email: setEmail,
    password: setPassword,
  };

  const isValidForm = email.isValid && password.isValid;

  const isFieldValid = (fieldName, value) => {
    const validators = {
      email: EMAIL_REGEX_KEY.test(String(value).toLowerCase()),
      password: value.length >= 6,
    };

    return validators[fieldName];
  };

  const onFieldChange = (fieldName, event) => {
    const input = getInput[fieldName];
    const isValid = isFieldValid(fieldName, event.target.value);
    getSetInput[fieldName]({
      ...input,
      value: event.target.value,
      isValid,
      showError: input.showError && !isValid,
    });
  };

  const validateInput = (fieldName) => {
    const input = getInput[fieldName];
    getSetInput[fieldName]({ ...input, showError: !input.isValid });
  };

  const inputErrorClass = (showError) => (showError ? 'b--red' : '');
  const disabledButtonClass = (isValidForm) => (isValidForm ? 'b bg-lightest-blue grow' : '');

  const onSignInSubmit = () => {
    if (email && password) {
      api.signInUser({ email, password }).then((user) => {
        if (user?.id) {
          loadUser(user);
          localStorage.setItem('user', JSON.stringify(user));
        } else {
          window.alert('Invalid user and/or password');
        }
      });
    } else {
      window.alert('Email and Password fields should be filled in');
    }
  };

  // const handleKeyPress = (event) => {
  //   return event.key === 'Enter' ? onSignInSubmit() : null;
  // };

  return (
    <form
      className="br3 ba b--black-10 mv4 w-100 w-50-m mw6 shadow-5 center form smaller"
      onSubmit={onSignInSubmit}
    >
      <main className="pa4 black-80">
        <div className="measure">
          <fieldset id="sign_in" className="ba b--transparent ph0 mh0">
            <h2 className="f2 ph0 ma0 noselect">Sign In</h2>
            <div>
              <label className="db fw6 lh-copy mv2" htmlFor="email">
                Email
                <input
                  className={`pa2 mb0 ba bg-white w-100 ${inputErrorClass(email.showError)}`}
                  type="email"
                  name="email"
                  id="email"
                  onChange={(event) => onFieldChange('email', event)}
                  onBlur={() => validateInput('email')}
                />
                {email.showError && (
                  <p className="red f6 absolute mv0 right-0 left-0">{email.errorText}</p>
                )}
              </label>
            </div>
            <div className="mt4">
              <label className="db fw6 lh-copy" htmlFor="password">
                Password
                <input
                  className={`pa2 mb0 ba bg-white w-100 ${inputErrorClass(password.showError)}`}
                  type="password"
                  name="password"
                  id="password"
                  onChange={(event) => onFieldChange('password', event)}
                  onBlur={() => validateInput('password')}
                />
                {password.showError && (
                  <p className="red f6 absolute mv0 right-0 left-0">{password.errorText}</p>
                )}
              </label>
            </div>
          </fieldset>
          <div className="flex flex-column items-center">
            <button
              className={`ph3 pv2 mt4 ba b--black f6 dib ${disabledButtonClass(isValidForm)}`}
              type="submit"
              disabled={!isValidForm}
            >
              Sign In
            </button>
            <button
              className="f5 dim black lh-copy mt3 b--none bg-transparent"
              type="button"
              onClick={() => history.push('/register')}
            >
              Register now, it's free!
            </button>
          </div>
        </div>
      </main>
    </form>
  );
};

SignIn.propTypes = {
  loadUser: PropTypes.func.isRequired,
};

export default SignIn;
