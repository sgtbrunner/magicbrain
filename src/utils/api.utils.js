const { REACT_APP_API_ENDPOINT } = process.env;

const fetchServer = ({
  path,
  method = 'POST',
  headers = { 'Content-Type': 'application/json' },
  body,
}) =>
  fetch(`${REACT_APP_API_ENDPOINT}${path}`, {
    method,
    headers,
    body: JSON.stringify({ ...body }),
  });

export default {
  signInUser: ({ email, password }) =>
    fetchServer({ path: '/signin', body: { email, password } }).then((response) => response.json()),

  registerUser: ({ name, email, password }) =>
    fetchServer({ path: '/register', body: { name, email, password } }).then((response) =>
      response.json()
    ),

  getImageBoundary: ({ input }) =>
    fetchServer({ path: '/imageurl', body: { input } }).then((response) => response.json()),

  updateImageCount: ({ id }) =>
    fetchServer({ path: '/image', method: 'PUT', body: { id } }).then((response) =>
      response.json()
    ),
};
