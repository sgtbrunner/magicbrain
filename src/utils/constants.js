export const PARTICLE_OPTIONS = {
  particles: {
    number: {
      value: 50,
      density: {
        enable: true,
        value_area: 500,
      },
    },
  },
};

export const { REACT_APP_API_ENDPOINT } = process.env;
