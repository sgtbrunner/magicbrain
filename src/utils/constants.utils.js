const PARTICLE_OPTIONS = {
  particles: {
    number: {
      value: 50,
      density: {
        enable: true,
        value_area: 500,
      },
    },
  },
};

export default PARTICLE_OPTIONS;
