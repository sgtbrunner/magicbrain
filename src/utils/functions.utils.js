export const getUserFromLocalStorage = () => localStorage.getItem('user');
