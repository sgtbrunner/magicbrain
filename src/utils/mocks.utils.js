export const USER = {
  id: 3,
  name: 'test',
  email: 'test@test.com',
  entries: 7,
};

export const OUTPUTS = {
  outputs: [
    {
      data: {
        regions: [
          {
            region_info: {
              bounding_box: {
                top_row: 0.5,
                left_col: 0.5,
                bottom_row: 0.5,
                right_col: 0.5,
              },
            },
          },
        ],
      },
    },
  ],
};

export const CREDENTIAL_ERROR = {
  error: 'You have entered an invalid username and/or password',
};

export const SERVER_ERROR = {
  error: 'Something went wrong. Please try again later.',
};
