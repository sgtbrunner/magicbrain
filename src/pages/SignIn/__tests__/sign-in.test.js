import { screen } from '@testing-library/react';

import { capitalizeFirstLetter } from '../../../utils/functions.utils';
import { mountComponent } from '../../../utils/test.utils';
import SignIn from '..';

const BUTTON = 'button';
const EMAIL = 'email';
const FORM_INPUT = 'form-input';
const HEADING = 'heading';
const PASSWORD = 'password';
const SIGN_IN = 'Sign in';

const mountSignInPage = () =>
  mountComponent({
    component: SignIn,
    props: {
      loadUser: jest.fn(),
    },
  });

describe('SignIn component', () => {
  it(`should match snapshot`, () => {
    const { container } = mountSignInPage();

    expect(container).toMatchSnapshot();
  });

  it(`should render 'Sign In' label and button`, () => {
    mountSignInPage();

    const label = screen.getByRole(HEADING, { name: SIGN_IN });
    const button = screen.getByRole(BUTTON, { name: SIGN_IN });

    expect(label).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  it(`should render two (2) input fields (email and password)`, () => {
    mountSignInPage();

    const inputFields = screen.getAllByTestId(FORM_INPUT);
    expect(inputFields).toHaveLength(2);

    const emailInput = screen.getByLabelText(capitalizeFirstLetter(EMAIL));
    expect(emailInput).toBeInTheDocument();

    const passwordInput = screen.getByLabelText(capitalizeFirstLetter(PASSWORD));
    expect(passwordInput).toBeInTheDocument();
  });

  it(`should have submit button disabled if input fields are invalid`, () => {
    mountSignInPage();

    const button = screen.getByRole(BUTTON, { name: SIGN_IN });
    expect(button).toBeDisabled();
  });

  it(`should render a register link`, () => {
    mountSignInPage();

    const linkButton = screen.getByRole(BUTTON, { name: `Register now, it's free!` });
    expect(linkButton).toBeInTheDocument();
  });
});
