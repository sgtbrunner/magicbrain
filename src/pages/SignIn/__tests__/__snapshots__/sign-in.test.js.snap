// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn page should match snapshot 1`] = `
<div>
  <form
    class="br3 ba b--black-10 mv4 mw6 shadow-5 center smaller"
    data-testid="user-form"
  >
    <main
      class="pv3 ph4 black-80"
    >
      <fieldset
        class="ba b--transparent ph0 mh0"
        id="sign-in"
      >
        <h2
          class="f2 ph0 ma0 noselect"
        >
          Sign in
        </h2>
        <div
          class=""
          data-testid="form-input"
        >
          <label
            class="db fw6 lh-copy mv2"
            for="email"
          >
            Email
            <input
              class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
              id="email"
              name="email"
              type="email"
            />
          </label>
        </div>
        <div
          class="mt4"
          data-testid="form-input"
        >
          <label
            class="db fw6 lh-copy mv2"
            for="password"
          >
            Password
            <input
              class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
              id="password"
              name="password"
              type="password"
            />
          </label>
        </div>
      </fieldset>
      <div
        class="flex flex-column items-center"
      >
        <button
          class="ph3 pv2 mt3 ba b--black f6 dib "
          disabled=""
          type="submit"
        >
          Sign in
        </button>
        <button
          class="f5 dim black lh-copy mt3 link b--none bg-transparent"
          type="button"
        >
          Register now, it's free!
        </button>
      </div>
    </main>
  </form>
</div>
`;
